File: backend/src/main/java/br/ufpr/sept/so2/shared/domain/AggregateRoot.java
package br.ufpr.sept.so2.shared.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public abstract class AggregateRoot {

    private final List<DomainEvent> domainEvents = new ArrayList<>();

    protected void registerEvent(DomainEvent event) {
        domainEvents.add(event);
    }

    public List<DomainEvent> getDomainEvents() {
        return Collections.unmodifiableList(domainEvents);
    }

    public List<DomainEvent> clearEvents() {
        List<DomainEvent> snapshot = List.copyOf(domainEvents);
        domainEvents.clear();
        return snapshot;
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/shared/domain/AggregateRoot.kt
package br.ufpr.sept.so2.shared.domain

abstract class AggregateRoot {
    private val domainEvents: MutableList<DomainEvent> = mutableListOf()

    protected fun registerEvent(event: DomainEvent) {
        domainEvents += event
    }

    fun getDomainEvents(): List<DomainEvent> = domainEvents.toList()

    fun clearEvents(): List<DomainEvent> {
        val snapshot = domainEvents.toList()
        domainEvents.clear()
        return snapshot
    }
}
